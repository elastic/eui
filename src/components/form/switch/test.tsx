import React from 'react';

import { EuiSwitch } from './switch';

export const Test = ({ checked }: { checked: boolean }) => (
  <EuiSwitch
    checked={checked}
    label="Test"
    onChange={e => console.log(e.target.checked)}
  />
);
