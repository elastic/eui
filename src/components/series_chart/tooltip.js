import React from 'react';
import _ from 'lodash';
import { Hint } from 'react-vis';
import PropTypes from 'prop-types';
// import {
//   colors,
//   unit,
//   units,
//   px,
//   borderRadius,
//   fontSize,
//   fontSizes
// } from '../../variables';
// import Legend from '../../Legend/Legend';

// const TooltipElm = styled.div`
//   margin: 0 ${px(unit)};
//   transform: translateY(-50%);
//   border: 1px solid ${colors.gray4};
//   background: ${colors.white};
//   border-radius: ${borderRadius};
//   font-size: ${fontSize};
//   color: ${colors.black};
// `;

// const Header = styled.div`
//   background: ${colors.gray5};
//   border-bottom: 1px solid ${colors.gray4};
//   border-radius: ${borderRadius} ${borderRadius} 0 0;
//   padding: ${px(units.half)};
//   color: ${colors.gray3};
// `;

// const Legends = styled.div`
//   display: flex;
//   flex-direction: column;
//   padding: ${px(units.half)};
//   padding: ${px(units.quarter)} ${px(unit)} ${px(units.quarter)}
//     ${px(units.half)};
//   font-size: ${fontSizes.small};
// `;

// const LegendContainer = styled.div`
//   display: flex;
//   align-items: center;
//   margin: ${px(units.quarter)} 0;
//   justify-content: space-between;
// `;

// const LegendGray = styled(Legend)`
//   color: ${colors.gray3};
// `;

// const Value = styled.div`
//   color: ${colors.gray2};
//   font-size: ${fontSize};
// `;

// <TooltipElm>
//     <Header>{header || moment(x).format('MMMM Do YYYY, HH:mm')}</Header>
//     <Legends>
//     {tooltipPoints.map((point, i) => (
//         <LegendContainer key={i}>
//         <LegendGray fontSize={fontSize.tiny} radius={units.half} color={point.color} text={point.text} />
//         <Value>{point.value}</Value>
//       </LegendContainer>
//     ))}
//   </Legends>
//   </TooltipElm>;

export default function Tooltip({ tooltipPoints, x, y, ...props }) {
  if (_.isEmpty(tooltipPoints)) {
    return null;
  }
  return <Hint {...props} value={{ x, y }} />;
}

Tooltip.propTypes = {
  header: PropTypes.string,
  tooltipPoints: PropTypes.array.isRequired,
  x: PropTypes.number,
  y: PropTypes.number
};

Tooltip.defaultProps = {};
