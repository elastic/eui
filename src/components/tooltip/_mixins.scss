@mixin euiTooltipSize($multiplier){
  max-width: $euiSize * $multiplier;

  .euiTooltip__content {
    white-space: inherit; // allow text wrapping of content
  }
}