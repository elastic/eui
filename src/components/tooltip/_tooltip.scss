.EuiTooltip {
  @include euiBottomShadow;
  @include euiFontSizeS();

  position: absolute; /* 1 */
  border-radius: $euiBorderRadius;
  padding: $euiSizeM;
  background-color: tintOrShade($euiColorFullShade, 25%, 90%);
  color: $euiColorGhost;
  max-width: 256px;
  opacity: 0;
  animation: EuiTooltipTop $euiAnimSpeedSlow ease-out $euiAnimSpeedNormal forwards;
  z-index: $euiZLevel9;

  &::before {
    content: "";
    position: absolute;
    bottom: -$euiSize/2;
    left: 50%;
    transform: translateX(-50%) rotateZ(45deg);
    transform-origin: center;
    background-color: tintOrShade($euiColorFullShade, 25%, 90%);
    width: $euiSize;
    height: $euiSize;
  }

  // Positions the arrow and animates in from the same side.
  &.EuiTooltip--right {
    animation-name: EuiTooltipRight;

    &:before {
      bottom: 50%;
      left: -$euiSize/2;
      transform: translateY(50%) rotateZ(45deg);
    }
  }

  &.EuiTooltip--bottom {
    animation-name: EuiTooltipBottom;

    &:before {
      bottom: auto;
      top: -$euiSize/2;
    }
  }

  &.EuiTooltip--left {
    animation-name: EuiTooltipLeft;

    &:before {
      bottom: 50%;
      left: auto;
      right: -$euiSize/2;
      transform: translateY(50%) rotateZ(45deg);
    }
  }

  .EuiTooltip__title {
    font-weight: $euiFontWeightBold;
    border-bottom: solid 1px tintOrShade($euiColorFullShade, 35%, 80%);
    padding-bottom: $euiSizeXS;
    margin-bottom: $euiSizeXS;
  }
}

.EuiTooltipAnchor {
  display: inline-block;
}

// Keyframes to animate in the tooltip.
@keyframes EuiTooltipTop {
  0% {
    opacity: 0;
    transform: translateY(-$euiSize);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

@keyframes EuiTooltipBottom {
  0% {
    opacity: 0;
    transform: translateY($euiSize);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

@keyframes EuiTooltipLeft {
  0% {
    opacity: 0;
    transform: translateX(-$euiSize);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

@keyframes EuiTooltipRight {
  0% {
    opacity: 0;
    transform: translateX($euiSize);
  }
  100% {
    opacity: 1;
    transform: translateY(0);
  }
}
