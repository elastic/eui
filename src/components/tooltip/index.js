export {
  EuiTooltip,
} from './tooltip';

export {
  EuiIconTip,
} from './icon_tip';
