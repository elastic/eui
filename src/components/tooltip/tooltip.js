import React, {
  Component,
  cloneElement,
  Fragment,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { EuiPortal } from '../portal';
import { EuiTooltipPopover } from './tooltip_popover';
import { calculatePopoverPosition } from '../../services';

import makeId from '../form/form_row/make_id';

const positionsToClassNameMap = {
  top: 'EuiTooltip--top',
  right: 'EuiTooltip--right',
  bottom: 'EuiTooltip--bottom',
  left: 'EuiTooltip--left',
};

export const POSITIONS = Object.keys(positionsToClassNameMap);

export class EuiTooltip extends Component {
  constructor(props) {
    super(props);

    this.state = {
      visible: false,
      hasFocus: false,
      calculatedPosition: this.props.position,
      toolTipStyles: {},
      id: this.props.id || makeId(),
    };
  }

  showToolTip = () => {
    this.setState({ visible: true });
  };

  positionToolTip = (toolTipBounds) => {
    const anchorBounds = this.anchor.getBoundingClientRect();
    const requestedPosition = this.props.position;

    const { position, left, top } = calculatePopoverPosition(anchorBounds, toolTipBounds, requestedPosition);

    const toolTipStyles = {
      top: top + window.scrollY,
      left,
    };

    this.setState({
      visible: true,
      calculatedPosition: position,
      toolTipStyles,
    });
  };

  hideToolTip = () => {
    this.setState({ visible: false });
  };

  onFocus = () => {
    this.setState({
      hasFocus: true,
    });
    this.showToolTip();
  };

  onBlur = () => {
    this.setState({
      hasFocus: false,
    });
    this.hideToolTip();
  };

  onMouseOut = (e) => {
    // Prevent mousing over children from hiding the tooltip by testing for whether the mouse has
    // left the anchor for a non-child.
    if (this.anchor === e.relatedTarget || !this.anchor.contains(e.relatedTarget)) {
      if (!this.state.hasFocus) {
        this.hideToolTip();
      }
    }
  };

  render() {
    const {
      children,
      className,
      anchorClassName,
      content,
      title,
      ...rest
    } = this.props;

    const classes = classNames(
      'EuiTooltip',
      positionsToClassNameMap[this.state.calculatedPosition],
      className
    );

    const anchorClasses = classNames(
      'EuiTooltipAnchor',
      anchorClassName,
    );

    let tooltip;
    if (this.state.visible) {
      tooltip = (
        <EuiPortal>
          <EuiTooltipPopover
            className={classes}
            style={this.state.toolTipStyles}
            positionToolTip={this.positionToolTip}
            title={title}
            id={this.state.id}
            role="tooltip"
            {...rest}
          >
            {content}
          </EuiTooltipPopover>
        </EuiPortal>
      );
    }

    const anchor = (
      <span
        ref={anchor => this.anchor = anchor}
        className={anchorClasses}
      >
        {cloneElement(children, {
          onFocus: this.showToolTip,
          onBlur: this.hideToolTip,
          'aria-describedby': this.state.id,
          onMouseOver: this.showToolTip,
          onMouseOut: this.onMouseOut
        })}
      </span>
    );

    return (
      <Fragment>
        {anchor}
        {tooltip}
      </Fragment>
    );
  }
}

EuiTooltip.propTypes = {
  /**
   * The in-view trigger for your tooltip.
   */
  children: PropTypes.element.isRequired,
  /**
   * The main content of your tooltip.
   */
  content: PropTypes.node.isRequired,

  /**
   * An optional title for your tooltip.
   */
  title: PropTypes.node,

  /**
   * Suggested position. If there is not enough room for it this will be changed.
   */
  position: PropTypes.oneOf(POSITIONS),

  /**
   * Passes onto the tooltip itself, not the trigger.
   */
  className: PropTypes.string,

  /**
   * Unless you provide one, this will be randomly generated.
   */
  id: PropTypes.string,
};

EuiTooltip.defaultProps = {
  position: 'top',
};
