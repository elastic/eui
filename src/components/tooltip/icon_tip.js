import React from 'react';
import PropTypes from 'prop-types';

import { EuiIcon } from '../icon';
import { EuiTooltip } from './tooltip';

export const EuiIconTip = ({ type, 'aria-label': ariaLabel, color, ...rest }) => (
  <EuiTooltip {...rest}>
    <EuiIcon tabIndex="0" type={type} color={color} aria-label={ariaLabel} />
  </EuiTooltip>
);

EuiIconTip.propTypes = {
  /**
   * The icon type.
   */
  type: PropTypes.string,

  /**
   * The icon color.
   */
  color: PropTypes.string,

  /**
   * Explain what this icon means for screen readers.
   */
  'aria-label': PropTypes.string,
};

EuiIconTip.defaultProps = {
  type: 'questionInCircle',
  'aria-label': 'Info',
};
