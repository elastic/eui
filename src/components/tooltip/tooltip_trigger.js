import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Tooltip } from './tooltip';
import { SIZE } from './tooltip_constants';

export class TooltipTrigger extends React.Component {
  static propTypes = {
    display: PropTypes.bool,
    title: PropTypes.string,
    tooltip: PropTypes.oneOfType([PropTypes.node, PropTypes.object]).isRequired,
    placement: PropTypes.oneOf(['left', 'right', 'bottom', 'top']),
    trigger: PropTypes.oneOf(['manual', 'hover', 'click']),
    clickHideDelay: PropTypes.number,
    onClick: PropTypes.func,
    onEntered: PropTypes.func,
    onExited: PropTypes.func,
    theme: PropTypes.oneOf(['dark', 'light']),
    size: PropTypes.oneOf([SIZE.AUTO, SIZE.SMALL, SIZE.MEDIUM, SIZE.LARGE]),
    isSticky: PropTypes.bool
  };

  static defaultProps = {
    display: false,
    placement: 'top',
    trigger: 'hover',
    clickHideDelay: 1000,
    onClick: () => {},
    onEntered: () => {},
    onExited: () => {},
    theme: 'dark',
    size: SIZE.AUTO,
    isSticky: false
  };

  constructor(props) {
    super(props);
    const openOnLoad = props.trigger === 'manual' ? props.display : false;
    this.state = {
      isVisible: openOnLoad,
      noOverflowPlacement: props.placement
    };
    this.clickHandler = this.clickHandler.bind(this);
  }

  noOverflowPlacement() {
    const domNode = ReactDOM.findDOMNode(this);
    const domNodeRect = domNode.getBoundingClientRect();
    const tooltipContainer = domNode.getElementsByClassName('tooltip-container')[0];
    const tooltipRect = tooltipContainer.getBoundingClientRect();
    const BUFFER = 10; // avoid tooltip from getting touching window edge and account for padding

    // determine tooltip overflow in each direction
    // negative values signal window overflow, large values signal lots of free space
    const tooltipOverflow = {
      top: domNodeRect.top - (tooltipRect.height + BUFFER),
      right: window.innerWidth - (domNodeRect.right + tooltipRect.width + BUFFER),
      bottom: window.innerHeight - (domNodeRect.bottom + tooltipRect.height + BUFFER),
      left: domNodeRect.left - (domNodeRect.width + BUFFER)
    };

    function hasCrossDimensionOverflow(key) {
      if (key === 'left' || key === 'right') {
        const domNodeCenterY = domNodeRect.top + (domNodeRect.height / 2);
        const tooltipTop = domNodeCenterY - ((tooltipRect.height / 2) + BUFFER);
        if (tooltipTop <= 0) {
          return true;
        }
        const tooltipBottom = domNodeCenterY + (tooltipRect.height / 2) + BUFFER;
        if (tooltipBottom >= window.innerHeight) {
          return true;
        }
      } else {
        const domNodeCenterX = domNodeRect.left + (domNodeRect.width / 2);
        const tooltipLeft = domNodeCenterX - ((tooltipRect.width / 2) + BUFFER);
        if (tooltipLeft <= 0) {
          return true;
        }
        const tooltipRight = domNodeCenterX + (tooltipRect.width / 2) + BUFFER;
        if (tooltipRight >= window.innerWidth) {
          return true;
        }
      }
      return false;
    }

    const userPlacement = this.props.placement;
    let bestPlacement = userPlacement;
    if (tooltipOverflow[userPlacement] <= 0 || hasCrossDimensionOverflow(userPlacement)) {
      // requested placement overflows window bounds
      // select direction what has the most free space
      Object.keys(tooltipOverflow).forEach((key) => {
        if (tooltipOverflow[key] > tooltipOverflow[bestPlacement] && !hasCrossDimensionOverflow(key)) {
          bestPlacement = key;
        }
      });
    }

    return bestPlacement;
  }

  hoverHandler(e) {
    this.setState({
      isVisible: e.type === 'mouseenter',
      noOverflowPlacement: this.noOverflowPlacement()
    });
  }

  clickHandler(e, onClick) {
    this.setState({
      isVisible: true,
      noOverflowPlacement: this.noOverflowPlacement()
    });
    onClick(e);
    setTimeout(() => {
      this.setState({ isVisible: false });
    }, this.props.clickHideDelay);
  }

  componentWillReceiveProps(nextProps) {
    const triggerChanged = this.props.trigger !== nextProps.trigger;
    const displayChanged = this.props.display !== nextProps.display;

    if (triggerChanged && nextProps.trigger === 'manual') {
      this.setState({ isVisible: nextProps.display });
    } else if (triggerChanged) {
      this.setState({ isVisible: false });
    } else if (displayChanged) {
      this.setState({ isVisible: nextProps.display });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if(prevState.isVisible && !this.state.isVisible) {
      this.props.onExited();
    } else if(!prevState.isVisible && this.state.isVisible) {
      this.props.onEntered();
    }
  }

  getTriggerHandler(trigger, onClick) {
    switch(trigger) {
      case 'click':
        return { onClick: e => this.clickHandler(e, onClick) };
      case 'manual':
        return {};
      default:
        return {
          onClick,
          onMouseEnter: this.hoverHandler.bind(this),
          onMouseLeave: this.hoverHandler.bind(this)
        };
    }
  }

  render() {
    const {
      isSticky,
      title,
      tooltip,
      trigger,
      className,
      clickHideDelay, // eslint-disable-line no-unused-vars
      onEntered, // eslint-disable-line no-unused-vars
      onExited, // eslint-disable-line no-unused-vars
      theme,
      size,
      onClick,
      display, // eslint-disable-line no-unused-vars
      ...rest
    } = this.props;
    const { isVisible } = this.state;

    const triggerHandler = this.getTriggerHandler(trigger, onClick);

    const newClasses = classnames('tooltip', className, {
      'tooltip-light': theme === 'light',
      [`tooltip-${this.state.noOverflowPlacement}`]: this.state.noOverflowPlacement !== 'top'
    });
    const newProps = {
      className: newClasses,
      ...triggerHandler,
      ...rest
    };
    const tooltipProps = { isSticky, size, isVisible, title };

    return (
      <div {...newProps}>
        {this.props.children}
        <Tooltip {...tooltipProps}>{tooltip}</Tooltip>
      </div>
    );
  }
}
