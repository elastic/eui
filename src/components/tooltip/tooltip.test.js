import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test';

import { EuiTooltip } from './tooltip';

describe('EuiTooltip', () => {
  test('is rendered', () => {
    const component = render(
      <EuiTooltip title="title" id="id" content="content" {...requiredProps}>
        <button>Trigger</button>
      </EuiTooltip>
    );

    expect(component)
      .toMatchSnapshot();
  });
});
