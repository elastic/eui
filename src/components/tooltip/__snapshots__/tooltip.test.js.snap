// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EuiTooltip is rendered 1`] = `
<span
  class="EuiTooltipAnchor"
>
  <button
    aria-describedby="id"
  >
    Trigger
  </button>
</span>
`;
